### ISSUE TITLE
twisted.python.modules.PythonPath.moduleLoader can repeatedly execute modules if they are imported from a buggy module

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3388]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3388|
|Type|defect|
|Created|2008-08-13 15:47:37Z|
|Branch|https://github.com/twisted/twisted/tree/module-loader-3388|

`moduleLoader` clears the module dictionary upon any exception.  This can cause modules to be executed repeatedly when they are imported as a side-effect of a failed import.  Since modules typically are not idempotent, this is a bad thing.

Consider:

```
exarkun@boson:~$ mkdir foo
exarkun@boson:~$ touch foo/__init__.py
exarkun@boson:~$ cat > foo/working.py
from twisted.internet.interfaces import IReactorTime
from twisted.python.components import registerAdapter
registerAdapter(int, int, IReactorTime)
exarkun@boson:~$ cat > foo/broken.py
import foo.working
1/0
exarkun@boson:~$ python -c 'from twisted.python.modules import getModule
from twisted.python import components # make sure the global adapter registry survives
from twisted.internet import interfaces # make sure IReactorTime survives 
for i in range(3):
    try:
        getModule("foo.broken").load() # this should fail the same way each time
    except ZeroDivisionError:
        pass
'

Traceback (most recent call last):
  File "<string>", line 6, in <module>
  File "/home/exarkun/Projects/Twisted/trunk/twisted/python/modules.py", line 380, in load
    return self.pathEntry.pythonPath.moduleLoader(self.name)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/python/modules.py", line 618, in moduleLoader
    return self._moduleLoader(modname)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/python/reflect.py", line 455, in namedAny
    topLevelPackage = _importAndCheckStack(trialname)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/python/reflect.py", line 391, in _importAndCheckStack
    return __import__(importName)
  File "foo/broken.py", line 1, in <module>
    import foo.working
  File "foo/working.py", line 3, in <module>
    registerAdapter(int, int, IReactorTime)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/python/components.py", line 90, in registerAdapter
    raise ValueError("an adapter (%s) was already registered." % (factory, ))
ValueError: an adapter (<type 'int'>) was already registered.
exarkun@boson:~$ 
```

It seems likely to me that the modules dictionary should be left alone.

<details><summary>Searchable metadata</summary>

```
trac-id__3388 3388
type__defect defect
reporter__exarkun exarkun
priority__high high
milestone__ 
branch__branches_module_loader_3388 branches/module-loader-3388
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1218642457000000 1218642457000000
changetime__1220287387000000 1220287387000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

Wow.  You're right.

The point of this behavior (as stated by the docstring) is to replicate the sys.modules preservation behavior present in python 2.4+ for python 2.3.  It clearly doesn't do that, since Python pays attention to exactly which modules import successfully.  Is this reasonably possible?  Should we even bother?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [24607]) Branching to 'module-loader-3388'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [24608]) replace moduleLoad with _moduleLoader

refs [#3388](https://github.com/twisted/twisted/issues/3388)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun removed owner|
|-|-|

Providing the 2.4+ behavior on 2.3 will be pretty hard.  I suspect it's possible, but I don't think it's worthwhile.  Code _is_ going to be using `import` _somewhere_ no matter what; that means that the possibility to have inconsistent modules in `sys.modules` if you're using Python 2.3 will exist, regardless of what `twisted.python.modules` does.

So I propose dropping the feature, which is what I've done in the branch.  There are no new tests because the change is a pure deletion.  I _think_ that means there's no reason to have new tests.  All the existing tests continue to pass.

I also changed some docstrings slightly, so this branch should also resolve [#3387](https://github.com/twisted/twisted/issues/3387).

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_6"></a>|@washort set owner to @exarkun|
|-|-|

Should attributes with leading underscores really be documented? It looks like the `sysPath` property is the one that should be accessed instead of `_sysPath`.

I guess the question is whether the docs are for someone who will be using the object or someone editing its definition. In the former case, I'd say it was correct as-is.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

All attributes should be documented.  So should all methods, classes, and modules.  The leading underscore tells everyone it's private, and the documentation tells people what it does/is/is for.  We don't need to pick one or the other. :)

`sysPath` has a docstring already, though I tweaked it to be a little more correct.

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_8"></a>|@washort commented|
|-|-|

OK, great. Looks ready to merge then.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @jml|
|-|-|

`twisted.trial.test.test_loader.PackageOrderingTest.test_sortPackagesDefaultOrder` fails on Python 2.3.

This seems to be because it does not expect to have error-holders for the failures in uberpackage.badpackage.test_module in the result.  Why doesn't it expect this, though?

I think it's probably a bug in the test.  On later versions of Python, it _does_ expect the error holders.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

I think this probably happens because with the previous behavior of `twisted.python.modules`, repeated calls to `modinfo.load(None)` would return `None` for `uberpackage.badpackage` and everything beneath it.  With the new behavior in this branch, the first call returns `None` but subsequent calls actually return a module.  This is, of course, the exact behavior which I deleted from `twisted.python.modules`, so that makes sense. :)

What is the behavior actually desired from trial here?  Is the repeated-loading error behavior actually important, or is an artifact of the test?  Can we just rewrite the test to build up the expected list in a different way?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun removed owner|
|-|-|

See r24661 for the resolution.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve set owner to @exarkun|
|-|-|

Fine, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set status to `closed`|
|-|-|

(In [24671]) Merge module-loader-3388

Author: exarkun
Reviewer: washort, therve
Fixes: [#3388](https://github.com/twisted/twisted/issues/3388)

Remove the error handling code from `twisted.python.modules` which
attempted to reset `sys.modules` to its pre-import value when an
import raised an exception.  This logic could lead to duplicate
instances of global state (one particular example being the running
processes state when parts of PyCrypto were not installed).  The
behavior only had value on Python 2.3, where Python itself did not
remove modules which failed to import with an error from the modules
dictionary.  `twisted.python.modules` will now behave more similarly
to `import` on all versions of Python.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
